cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java

```diff
@@ -336,9 +337,110 @@ public void rollback(TopicId topicId, RollbackDetail rollbackDetail)
     throw new IOException("NOT IMPLEMENTED");
   }
 
+  /**
+   * Fetches messages from the specified topic.
+   *
+   * <p>The fetch operation utilizes the following protocol:
+   *     <ul>
+   *         <li>**MessageId Construction:**
+   *             MessageId is constructed as the concatenation of `sequence_id` and `publish_ts`.
+   *             This guarantees the following: If the `publish_ts` of message A is less than the
+   *            `publish_ts` of message B, then message A is delivered before message B.
+   *         </li>
+   *         <li>**Fetch Logic:**
+   *             - If the provided `messageId` exists in-memory, the iterator pointing to the next message is returned.
+   *             - Otherwise, a Spanner read operation is issued to fetch messages
+   *               with `publish_ts` greater than the `publish_ts` in the provided `messageId`.
+   *             - Upon reading rows, chunked messages are reconstructed as a single message and are held in memory.
+   *         </li>
+   *     </ul>
+   * </p>
+   */
   @Override
   public CloseableIterator<RawMessage> fetch(MessageFetchRequest messageFetchRequest)
       throws TopicNotFoundException, IOException {
-    throw new IOException("NOT IMPLEMENTED");
+    // If the client is fetching for the first time ever,
+    // they should fetch data for publish_ts > 0 and sequence_id > -1.
+    Long startTime = 0L;
+    if (messageFetchRequest.getStartTime() != null) {
+      startTime = messageFetchRequest.getStartTime();
+    }
+    short sequenceId = -1;
+    // MessageId is constructed as the concatenation of SequenceNo and CommitTs columns.
+    // These fields are set in the offset field of the message fetch request.
+    byte[] id = messageFetchRequest.getStartOffset();
+    if (id != null) {
+      int offset = 0;
+      startTime = Bytes.toLong(id, offset);
+      offset += Bytes.SIZEOF_LONG;
+      sequenceId = Bytes.toShort(id, offset);
+    }
+
+    LOG.trace("Fetching message for topic from : {} with start time : {} sequenceId : {}",
```

Comment:
nit: rename to `Fetching message from {} topic with timestamp : {} , and  sequence Id : {}`